tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/routes/DownloadAllRoutes.scala

```diff
@@ -0,0 +1,190 @@
+      .flatMap(attachment => BlobId.of(attachment.getAttachmentId.getId)
+        .fold(e => SMono.error(e),
+          blobResolvers.resolve(_, mailboxSession)))
+      .collectSeq()
+      .flatMap(blobs => downloadBlob(
+        optionalName = queryParam(request, nameParam),
+        response = response,
+        blobContentType = ContentType.of("zip"),
+        blobs)
+        .`then`())
+
+  private def getIfOwner(request: HttpServerRequest, response: HttpServerResponse, mailboxSession: MailboxSession): SMono[Unit] =
+    Id.validate(request.param(accountIdParam)) match {
+      case Right(id: Id) => sessionTranslator.delegateIfNeeded(mailboxSession, AccountId(id))
+        .flatMap(session => get(request, response, session))
+      case Left(throwable: Throwable) => SMono.error(throwable)
+    }
+
+  private def downloadBlob(optionalName: Option[String],
+                           response: HttpServerResponse,
+                           blobContentType: ContentType,
+                           blobs: Seq[Blob]): SMono[Unit] = {
+    val resourceSupplier: Callable[Seq[Blob]] = () => blobs
+    val sourceSupplier: java.util.function.Function[Seq[Blob], Mono[Void]] = newBlobs => SMono(addContentDispositionHeader(optionalName)
+      .compose(addCacheControlHeader())
+      .apply(response)
+      .header(CONTENT_TYPE, sanitizeHeaderValue(blobContentType.asString))
+      .status(OK)
+      .send(ZipUtil.createZipStream(newBlobs.asJavaCollection)
+        .map(Unpooled.wrappedBuffer(_))
+        .subscribeOn(Schedulers.boundedElastic()))).asJava()
+    val resourceRelease: Consumer[Seq[Blob]] = (blobs: Seq[Blob]) => blobs.map(blob => blob.content.close())
+
+    SMono.fromPublisher(Mono.using(
+        resourceSupplier,
+        sourceSupplier,
+        resourceRelease))
+      .`then`
+  }
+
+  private def addContentDispositionHeader(optionalName: Option[String]): HttpServerResponse => HttpServerResponse =
+    resp => optionalName.map(addContentDispositionHeaderRegardingEncoding(_, resp))
+      .getOrElse(resp)
+
+  private def sanitizeHeaderValue(s: String): String =
+    if (HttpHeaderValidationUtil.validateValidHeaderValue(s) == -1) {
+      s
+    } else {
+      "application/octet-stream"
+    }
+
+  private def addContentLengthHeader(sizeTry: Try[Size]): HttpServerResponse => HttpServerResponse =
+    resp => sizeTry
+      .map(size => resp.header("Content-Length", size.value.toString))
+      .getOrElse(resp)
+
+  private def addCacheControlHeader(): HttpServerResponse => HttpServerResponse =
+    resp => resp.header(HttpHeaderNames.CACHE_CONTROL, "private, immutable, max-age=31536000")
+
+  private def addContentDispositionHeaderRegardingEncoding(name: String, resp: HttpServerResponse): HttpServerResponse =
+    if (CharMatcher.ascii.matchesAllOf(name)) {
+      Try(resp.header("Content-Disposition", "attachment; filename=\"" + name + "\""))
```

Comment:
I can trivially inject HTTP headers using values like `abc.zip"/nEvil: PWND` or at the very list pollute the HTTP ContentDisposition parameters: `abc.zip" otherParam="PWND"`

We shall beware of String concat.

How does the download route avoids those pitfalls?

Reply:
> I can trivially inject HTTP headers using values like abc.zip"/nEvil: PWND or at the very list pollute the HTTP ContentDisposition parameters: abc.zip" otherParam="PWND"
> 
> We shall beware of String concat.
>
> How does the download route avoids those pitfalls?


This comment in neither handled nor answered.